UnityProject/Assets/Scripts/PlayGroups/PlayerMove.cs

```diff
@@ -193,69 +193,79 @@ private Vector3Int GetMoveDirection(KeyCode action)
 		/// <summary>
 		///     Check current and next tiles to determine their status and if movement is allowed
 		/// </summary>
-		private Vector3Int AdjustDirection(Vector3Int currentPosition, Vector3Int direction)
+		private Vector3Int AdjustDirection(Vector3Int currentPosition, Vector3Int direction, bool isReplay)
 		{
 			if (isGhost) {
 				return direction;
 			}
 
-			//Is the current tile restrictive?
 			Vector3Int newPos = currentPosition + direction;
-			Vector3 newRayPos = transform.position + direction;
-			rayHit = Physics2D.RaycastAll(newRayPos, (Vector3)direction, 0.2f, hitCheckLayers);
-			Debug.DrawLine(newRayPos, newRayPos + ((Vector3)direction * 0.2f), Color.red, 1f);
-			for (int i = 0; i < rayHit.Length; i++) {
-				//checks to see if the matrix has changed
-				if (rayHit[i].collider.gameObject.layer == 24
-				   && rayHit[i].collider != curMatrixCol) {
-					curMatrixCol = rayHit[i].collider;
-					ChangeMatricies(rayHit[i].collider.gameObject.transform.parent);
-					Debug.Log("Change Matricies");
-				}
 
-				//Detected windows or walls (global will detect on other matricies also)
-				if (rayHit[i].collider.gameObject.layer == 9
-				   || rayHit[i].collider.gameObject.layer == 18) {
-					return Vector3Int.zero;
-				}
+			//isReplay tells AdjustDirection if the move being carried out is a replay move for prediction or not
+			//a replay move is a move that has already been carried out on the LocalPlayer's client
+			if (!isReplay) {
+				
+				Vector3 newRayPos = transform.position + direction;
+				rayHit = Physics2D.RaycastAll(newRayPos, (Vector3)direction, 0.2f, hitCheckLayers);
+			//	Debug.DrawLine(newRayPos, newRayPos + ((Vector3)direction * 0.2f), Color.red, 1f);
+
+				//Detect new matrices
+				for (int i = 0; i < rayHit.Length; i++) {
+					//checks to see if the matrix has changed
+					if (rayHit[i].collider.gameObject.layer == 24
+						&& rayHit[i].collider != curMatrixCol) {
+						curMatrixCol = rayHit[i].collider;
+						ChangeMatricies(rayHit[i].collider.gameObject.transform.parent);
+						Debug.Log("Change Matricies");
+					}
 
-				//Door closed layer
-				if (rayHit[i].collider.gameObject.layer == 17) {
-					DoorController doorController = rayHit[i].collider.gameObject.GetComponent<DoorController>();
+					//Detected windows or walls across matrices or from space:
+					if (rayHit[i].collider.gameObject.layer == 9
+					   || rayHit[i].collider.gameObject.layer == 18) {
+						return Vector3Int.zero;
+					}
 
-					// Attempt to open door that could be on another layer
-					if (doorController != null && allowInput) {
-						pna.CmdCheckDoorPermissions(doorController.gameObject, gameObject);
-						allowInput = false;
-						StartCoroutine(DoorInputCoolDown());
+					//Door closed layer (matrix independent)
+					if (rayHit[i].collider.gameObject.layer == 17) {
+						DoorController doorController = rayHit[i].collider.gameObject.GetComponent<DoorController>();
```

Comment:
Are we going to list all possible interactions here? (like, what if there is a pushable on the matrix border?) Sounds like a potential maintenance hell

Reply:
It was a rough draft, trying to detect things on different layers. We def need a better solution and soon